Interactive Environment container waiting and swarm management

natefoo opened the pull request:
1. Create a new API endpoint and client spin to wait for GIE containers to become ready. IEs are not required to use it, but usage does require a slight change to their `load_notebook()` functions.

2. Add a Docker swarm manager daemon that runs automatically to add nodes when services are waiting, clean up exited services, and remove idle nodes. Shuts down automatically when there's nothing to do.

The format of the GIE proxy session database has changed, if GIEs are running when you restart Galaxy, they will be interrupted. If you run the GIE proxy separate from Galaxy, you should restart it after upgrading.

The cron job to clean services has been removed since it is obviated by the swarm manager.

I am not excited about the placement of the swarm manager code, but couldn't come up with a better location for it. Ideas welcome.

natefoo committed:
GIE container waiting and swarm management.

1. Create a new API endpoint and client spin to wait for GIE containers to
   become ready. IEs are not required to use it, but usage does require
   a slight change to their `load_notebook()` functions.

2. Add a Docker swarm manager daemon that runs automatically
   to add nodes when services are waiting, clean up exited services, and
   remove idle nodes. Shuts down automatically when there's nothing to
   do.

The format of the GIE proxy session database has changed, if GIEs are
running when you restart Galaxy, they will be interrupted. If you run
the GIE proxy separate from Galaxy, you should restart it after
upgrading.

martenson commented on lib/galaxy/webapps/galaxy/buildapp.py:
```diff
@@ -256,6 +256,7 @@ def populate_api_routes( webapp, app ):
     webapp.mapper.resource( 'remote_file', 'remote_files', path_prefix='/api' )
     webapp.mapper.resource( 'group', 'groups', path_prefix='/api' )
     webapp.mapper.resource_with_deleted( 'quota', 'quotas', path_prefix='/api' )
+    webapp.mapper.connect( '/api/interactive_environment_ready', controller='interactive_environment_ready' )
```
fwiw this is not very RESTful

natefoo replied:
Yeah... suggestions welcome.

martenson replied:
Should this be part of the `/api/user`?

dannon replied:
I know they're launched *from* dataset(hda)->visualizations, but are they actually tied to them?  Since they ultimately end up being a history item, it feels like history/contents/<whatever_it_is> should show status.

martenson replied:
@dannon iiuc this API is tied to user's session - not to a particular GIE

natefoo replied:
Eventually we did want to have the ability to run multiple concurrent GIEs. Only one per session right now, though.

dannon replied:
Planning for the future (since this is the API and all), maybe go with something ```GET /api/interactive_environment/<key>/ready```?

jmchilton replied:
I'd consider a web controller for this instead of an API controller - it is what I think we use for something like getting the current history ID that isn't stateless or RESTful.

martenson reviewed (approved):
LGTM. I would still like to see the API under `/api/user` or better.

natefoo commented:
Added some changes with a rebase. Moved `/api/interactive_environment_ready` to a regular controller at `/interactive_environments/ready` and moved the container package to the top level of `galaxy`, plus a couple small bug fixes.

jmchilton commented:
Cool stuff - thanks!
